File: src/multicorpus_engine/stats.py
"""Lexical statistics engine.

Computes word frequency distributions over a configurable sub-corpus
(slot).  The slot is defined by optional filters on document metadata
(doc_ids, language, doc_role, resource_type, family_id).

API:
    compute_lexical_stats(conn, slot) -> StatsResult
    compute_stats_compare(conn, slot_a, slot_b) -> StatsCompareResult
"""

from __future__ import annotations

import re
import sqlite3
from collections import Counter
from dataclasses import dataclass, field
from typing import Any, Optional

_WORD_RE = re.compile(r"[^\W\d_]+", re.UNICODE)


# ─── Data classes ─────────────────────────────────────────────────────────────

@dataclass
class StatsSlot:
    """Defines a sub-corpus over which to compute statistics."""
    doc_ids: Optional[list[int]] = None       # None = all documents
    language: Optional[str] = None
    doc_role: Optional[str] = None
    resource_type: Optional[str] = None
    family_id: Optional[int] = None           # include parent + children
    top_n: int = 50
    min_length: int = 2


@dataclass
class StatsWord:
    word: str
    count: int
    freq_pct: float   # percentage of total tokens


@dataclass
class StatsResult:
    label: str
    total_tokens: int
    vocabulary_size: int
    total_units: int
    total_docs: int
    avg_tokens_per_unit: float
    top_words: list[StatsWord]
    rare_words: list[StatsWord]


@dataclass
class StatsCompareWord:
    word: str
    count_a: int
    count_b: int
    freq_a: float   # % of corpus A
    freq_b: float   # % of corpus B
    ratio: float    # freq_a / freq_b (0 if freq_b == 0; -1 means only in B)


@dataclass
class StatsCompareResult:
    label_a: str
    label_b: str
    summary_a: StatsResult
    summary_b: StatsResult
    comparison: list[StatsCompareWord]   # union of top words from A and B


# ─── Helpers ──────────────────────────────────────────────────────────────────

def _tokenize(text: str, min_length: int = 2) -> list[str]:
    return [w.lower() for w in _WORD_RE.findall(text) if len(w) >= min_length]


def _slot_to_doc_ids(conn: sqlite3.Connection, slot: StatsSlot) -> list[int] | None:
    """Resolve slot filters to a concrete list of doc_ids (or None = all)."""
    # Family expansion: parent + all children
    if slot.family_id is not None:
        child_rows = conn.execute(
            "SELECT doc_id FROM doc_relations"
            " WHERE target_doc_id = ? AND relation_type IN ('translation_of', 'excerpt_of')",
            [slot.family_id],
        ).fetchall()
        family_ids = [slot.family_id] + [r[0] for r in child_rows]
        # Intersect with any explicit doc_ids filter
        if slot.doc_ids is not None:
            return [d for d in family_ids if d in set(slot.doc_ids)]
        return family_ids

    return slot.doc_ids  # may be None (= all)


def _fetch_texts(conn: sqlite3.Connection, slot: StatsSlot) -> tuple[list[str], int, int]:
    """Fetch text_norm for units matching the slot.

    Returns (texts, unit_count, doc_count).
    """
    effective_doc_ids = _slot_to_doc_ids(conn, slot)

    filters: list[str] = ["u.unit_type = 'line'"]
    params: list[Any] = []

    if slot.language:
        filters.append("d.language = ?")
        params.append(slot.language)
    if slot.doc_role:
        filters.append("d.doc_role = ?")
        params.append(slot.doc_role)
    if slot.resource_type:
        filters.append("d.resource_type = ?")
        params.append(slot.resource_type)
    if effective_doc_ids is not None:
        if not effective_doc_ids:
            return [], 0, 0
        placeholders = ",".join("?" * len(effective_doc_ids))
        filters.append(f"u.doc_id IN ({placeholders})")
        params.extend(effective_doc_ids)

    where = " AND ".join(filters)
    rows = conn.execute(
        f"""
        SELECT u.text_norm, u.doc_id
        FROM units u
        JOIN documents d ON d.doc_id = u.doc_id
        WHERE {where}
        ORDER BY u.doc_id, u.n
        """,
        params,
    ).fetchall()

    texts = [row[0] or "" for row in rows]
    doc_count = len({row[1] for row in rows})
    return texts, len(texts), doc_count


def _compute_from_texts(
    texts: list[str],
    unit_count: int,
    doc_count: int,
    label: str,
    top_n: int,
    min_length: int,
) -> StatsResult:
    counter: Counter[str] = Counter()
    for text in texts:
        counter.update(_tokenize(text, min_length))

    total_tokens = sum(counter.values())
    vocabulary_size = len(counter)
    avg = total_tokens / unit_count if unit_count else 0.0

    def _pct(count: int) -> float:
        return round(count / total_tokens * 100, 3) if total_tokens else 0.0

    top_words = [
        StatsWord(word=w, count=c, freq_pct=_pct(c))
        for w, c in counter.most_common(top_n)
    ]
    rare_words = [
        StatsWord(word=w, count=c, freq_pct=_pct(c))
        for w, c in reversed(counter.most_common())
    ][:top_n]

    return StatsResult(
        label=label,
        total_tokens=total_tokens,
        vocabulary_size=vocabulary_size,
        total_units=unit_count,
        total_docs=doc_count,
        avg_tokens_per_unit=round(avg, 2),
        top_words=top_words,
        rare_words=rare_words,
    )


# ─── Public API ───────────────────────────────────────────────────────────────

def compute_lexical_stats(
    conn: sqlite3.Connection,
    slot: StatsSlot,
    label: str = "",
) -> StatsResult:
    """Compute lexical frequency statistics for a sub-corpus slot."""
    texts, unit_count, doc_count = _fetch_texts(conn, slot)
    return _compute_from_texts(texts, unit_count, doc_count, label, slot.top_n, slot.min_length)


def compute_stats_compare(
    conn: sqlite3.Connection,
    slot_a: StatsSlot,
    slot_b: StatsSlot,
    label_a: str = "A",
    label_b: str = "B",
) -> StatsCompareResult:
    """Compare frequency distributions of two sub-corpus slots."""
    texts_a, unit_a, doc_a = _fetch_texts(conn, slot_a)
    texts_b, unit_b, doc_b = _fetch_texts(conn, slot_b)

    summary_a = _compute_from_texts(texts_a, unit_a, doc_a, label_a, slot_a.top_n, slot_a.min_length)
    summary_b = _compute_from_texts(texts_b, unit_b, doc_b, label_b, slot_b.top_n, slot_b.min_length)

    # Union of words appearing in top or rare of either side
    words_a = {w.word: w.count for w in summary_a.top_words + summary_a.rare_words}
    words_b = {w.word: w.count for w in summary_b.top_words + summary_b.rare_words}
    all_words = sorted(set(words_a) | set(words_b))

    total_a = summary_a.total_tokens or 1
    total_b = summary_b.total_tokens or 1

    comparison: list[StatsCompareWord] = []
    for word in all_words:
        ca = words_a.get(word, 0)
        cb = words_b.get(word, 0)
        fa = round(ca / total_a * 100, 4)
        fb = round(cb / total_b * 100, 4)
        if fb == 0:
            ratio = 0.0 if ca == 0 else float("inf")
        else:
            ratio = round(fa / fb, 3)
        comparison.append(StatsCompareWord(
            word=word, count_a=ca, count_b=cb,
            freq_a=fa, freq_b=fb, ratio=ratio,
        ))

    # Sort by sum of frequencies descending
    comparison.sort(key=lambda x: x.freq_a + x.freq_b, reverse=True)

    return StatsCompareResult(
        label_a=label_a,
        label_b=label_b,
        summary_a=summary_a,
        summary_b=summary_b,
        comparison=comparison,
    )


# ─── Serialisation ────────────────────────────────────────────────────────────

def stats_result_to_dict(r: StatsResult) -> dict:
    return {
        "label": r.label,
        "total_tokens": r.total_tokens,
        "vocabulary_size": r.vocabulary_size,
        "total_units": r.total_units,
        "total_docs": r.total_docs,
        "avg_tokens_per_unit": r.avg_tokens_per_unit,
        "top_words": [{"word": w.word, "count": w.count, "freq_pct": w.freq_pct} for w in r.top_words],
        "rare_words": [{"word": w.word, "count": w.count, "freq_pct": w.freq_pct} for w in r.rare_words],
    }


def stats_compare_result_to_dict(r: StatsCompareResult) -> dict:
    return {
        "label_a": r.label_a,
        "label_b": r.label_b,
        "summary_a": stats_result_to_dict(r.summary_a),
        "summary_b": stats_result_to_dict(r.summary_b),
        "comparison": [
            {
                "word": c.word,
                "count_a": c.count_a,
                "count_b": c.count_b,
                "freq_a": c.freq_a,
                "freq_b": c.freq_b,
                "ratio": c.ratio if c.ratio != float("inf") else None,
            }
            for c in r.comparison
        ],
    }

File: tauri-app/src/features/stats.ts
/**
 * features/stats.ts — Lexical statistics panel for the concordancer.
 *
 * Two modes:
 *   Simple  — frequency list for one sub-corpus slot.
 *   Compare — side-by-side A/B comparison of two slots.
 *
 * Each slot is defined by optional filters: doc_ids, language, doc_role,
 * resource_type, family_id, top_n.
 *
 * Inspired by HIMYC-Tauri's concordancierModule stats panel.
 */

import { state } from "../state";
import { elt } from "../ui/dom";
import {
  fetchLexicalStats, fetchStatsCompare,
  type StatsSlot, type StatsResult, type StatsCompareResult, type StatsWord, type StatsCompareWord,
} from "../lib/sidecarClient";

// ─── Panel state ──────────────────────────────────────────────────────────────

let _compareMode = false;
let _lastResult: StatsResult | null = null;
let _lastCompare: StatsCompareResult | null = null;

// ─── Helpers ──────────────────────────────────────────────────────────────────

function _readSlot(prefix: string): StatsSlot {
  const docIdsRaw = (document.getElementById(`stats-${prefix}-doc-ids`) as HTMLInputElement | null)?.value.trim();
  const docIds = docIdsRaw
    ? docIdsRaw.split(",").map(s => parseInt(s.trim(), 10)).filter(n => !isNaN(n))
    : null;

  const language = (document.getElementById(`stats-${prefix}-lang`) as HTMLSelectElement | null)?.value || null;
  const docRole = (document.getElementById(`stats-${prefix}-role`) as HTMLSelectElement | null)?.value || null;
  const resourceType = (document.getElementById(`stats-${prefix}-restype`) as HTMLSelectElement | null)?.value || null;
  const familyIdRaw = (document.getElementById(`stats-${prefix}-family`) as HTMLSelectElement | null)?.value;
  const familyId = familyIdRaw ? parseInt(familyIdRaw, 10) : null;
  const topN = parseInt((document.getElementById("stats-top-n") as HTMLInputElement | null)?.value ?? "50", 10);

  return {
    doc_ids: docIds && docIds.length > 0 ? docIds : null,
    language,
    doc_role: docRole,
    resource_type: resourceType,
    family_id: isNaN(familyId as number) ? null : familyId,
    top_n: Math.max(5, Math.min(500, topN || 50)),
  };
}

function _readLabel(prefix: string): string {
  return (document.getElementById(`stats-${prefix}-label`) as HTMLInputElement | null)?.value.trim() || (prefix === "a" ? "Corpus A" : "Corpus B");
}

function _escHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function _barHtml(pct: number, cls = "stats-bar-a"): string {
  const w = Math.min(100, pct * 5).toFixed(1); // scale: 20% = full bar
  return `<span class="stats-bar-wrap"><span class="${cls}" style="width:${w}%"></span></span>`;
}

// ─── Populate filter selects ──────────────────────────────────────────────────

export function populateStatsSelects(): void {
  ["a", "b"].forEach(prefix => {
    const langSel = document.getElementById(`stats-${prefix}-lang`) as HTMLSelectElement | null;
    const roleSel = document.getElementById(`stats-${prefix}-role`) as HTMLSelectElement | null;
    const restypeSel = document.getElementById(`stats-${prefix}-restype`) as HTMLSelectElement | null;
    const famSel = document.getElementById(`stats-${prefix}-family`) as HTMLSelectElement | null;

    if (langSel) {
      const langs = [...new Set(state.docs.map(d => d.language).filter(Boolean))].sort();
      langSel.innerHTML = `<option value="">Toutes</option>` +
        langs.map(l => `<option value="${_escHtml(l)}">${_escHtml(l)}</option>`).join("");
    }
    if (roleSel) {
      const roles = [...new Set(state.docs.map(d => d.doc_role).filter(Boolean) as string[])].sort();
      roleSel.innerHTML = `<option value="">Tous</option>` +
        roles.map(r => `<option value="${_escHtml(r)}">${_escHtml(r)}</option>`).join("");
    }
    if (restypeSel) {
      const types = [...new Set(state.docs.map(d => d.resource_type).filter(Boolean) as string[])].sort();
      restypeSel.innerHTML = `<option value="">Tous</option>` +
        types.map(t => `<option value="${_escHtml(t)}">${_escHtml(t)}</option>`).join("");
    }
    if (famSel) {
      famSel.innerHTML = `<option value="">Toutes familles</option>` +
        state.families.map(f => {
          const label = f.parent?.title ?? `Famille #${f.family_id}`;
          return `<option value="${f.family_id}">${_escHtml(label)}</option>`;
        }).join("");
    }
  });
}

// ─── Slot B visibility ────────────────────────────────────────────────────────

function _updateCompareModeUI(): void {
  const slotB = document.getElementById("stats-slot-b");
  const compareCb = document.getElementById("stats-compare-cb") as HTMLInputElement | null;
  _compareMode = compareCb?.checked ?? false;
  if (slotB) slotB.style.display = _compareMode ? "" : "none";
}

// ─── Rendering ────────────────────────────────────────────────────────────────

function _renderKpiCards(r: StatsResult): string {
  const kpis = [
    { label: "Tokens total", value: r.total_tokens.toLocaleString("fr-FR") },
    { label: "Vocabulaire", value: r.vocabulary_size.toLocaleString("fr-FR") },
    { label: "Unités", value: r.total_units.toLocaleString("fr-FR") },
    { label: "Documents", value: r.total_docs.toLocaleString("fr-FR") },
    { label: "Moy. tokens/unité", value: r.avg_tokens_per_unit.toFixed(1) },
  ];
  return `<div class="stats-kpi-row">${kpis.map(k =>
    `<div class="stats-kpi-card"><div class="stats-kpi-value">${k.value}</div><div class="stats-kpi-label">${k.label}</div></div>`
  ).join("")}</div>`;
}

function _renderWordTable(words: StatsWord[], title: string, exportId: string): string {
  const rows = words.map(w =>
    `<tr>
      <td class="stats-word">${_escHtml(w.word)}</td>
      <td class="stats-count">${w.count}</td>
      <td class="stats-pct">${w.freq_pct.toFixed(3)}&nbsp;%</td>
      <td class="stats-bar-cell">${_barHtml(w.freq_pct)}</td>
    </tr>`
  ).join("");
  return `
    <div class="stats-table-head">
      <strong>${title}</strong>
      <button class="stats-export-btn" data-export="${exportId}" type="button">⬇ CSV</button>
    </div>
    <div class="stats-table-wrap">
      <table class="stats-table">
        <thead><tr><th>Mot</th><th>Occurrences</th><th>Fréquence</th><th></th></tr></thead>
        <tbody>${rows}</tbody>
      </table>
    </div>`;
}

function _renderSingleStats(r: StatsResult): string {
  const label = r.label ? `<h3 class="stats-section-title">${_escHtml(r.label)}</h3>` : "";
  return label
    + _renderKpiCards(r)
    + _renderWordTable(r.top_words, `Top ${r.top_words.length} mots les plus fréquents`, "top")
    + _renderWordTable(r.rare_words, `Top ${r.rare_words.length} mots les plus rares`, "rare");
}

function _renderCompareStats(r: StatsCompareResult): string {
  const summaryHtml = `
    <div class="stats-compare-summaries">
      <div class="stats-compare-col stats-compare-a">
        <div class="stats-compare-col-head">${_escHtml(r.label_a)}</div>
        ${_renderKpiCards(r.summary_a)}
      </div>
      <div class="stats-compare-col stats-compare-b">
        <div class="stats-compare-col-head">${_escHtml(r.label_b)}</div>
        ${_renderKpiCards(r.summary_b)}
      </div>
    </div>`;

  const rows = r.comparison.map(w => {
    const ratioDisplay = w.ratio === null ? "∞" : w.ratio === 0 ? "—" : w.ratio.toFixed(2);
    const ratioClass = w.ratio === null ? "stats-ratio-inf" : (w.ratio ?? 0) >= 1 ? "stats-ratio-a" : "stats-ratio-b";
    return `<tr>
      <td class="stats-word">${_escHtml(w.word)}</td>
      <td class="stats-count">${w.count_a}</td>
      <td class="stats-pct">${w.freq_a.toFixed(3)}&nbsp;%</td>
      <td class="stats-bar-cell">${_barHtml(w.freq_a, "stats-bar-a")}</td>
      <td class="stats-count">${w.count_b}</td>
      <td class="stats-pct">${w.freq_b.toFixed(3)}&nbsp;%</td>
      <td class="stats-bar-cell">${_barHtml(w.freq_b, "stats-bar-b")}</td>
      <td class="stats-ratio ${ratioClass}">${ratioDisplay}</td>
    </tr>`;
  }).join("");

  const compareTable = `
    <div class="stats-table-head">
      <strong>Comparaison (${r.comparison.length} mots)</strong>
      <button class="stats-export-btn" data-export="compare" type="button">⬇ CSV</button>
    </div>
    <div class="stats-table-wrap">
      <table class="stats-table stats-compare-table">
        <thead>
          <tr>
            <th rowspan="2">Mot</th>
            <th colspan="3" class="stats-col-head-a">${_escHtml(r.label_a)}</th>
            <th colspan="3" class="stats-col-head-b">${_escHtml(r.label_b)}</th>
            <th rowspan="2">Ratio</th>
          </tr>
          <tr>
            <th>N</th><th>Freq.&nbsp;%</th><th></th>
            <th>N</th><th>Freq.&nbsp;%</th><th></th>
          </tr>
        </thead>
        <tbody>${rows}</tbody>
      </table>
    </div>`;

  return summaryHtml + compareTable;
}

// ─── CSV export ───────────────────────────────────────────────────────────────

function _exportCsv(data: string, filename: string): void {
  const blob = new Blob([data], { type: "text/csv;charset=utf-8;" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

function _handleExport(exportType: string): void {
  if (!_compareMode && _lastResult) {
    const r = _lastResult;
    if (exportType === "top") {
      const csv = "mot,occurrences,freq_pct\n" + r.top_words.map(w => `${w.word},${w.count},${w.freq_pct}`).join("\n");
      _exportCsv(csv, `stats_top_${r.label || "corpus"}.csv`);
    } else if (exportType === "rare") {
      const csv = "mot,occurrences,freq_pct\n" + r.rare_words.map(w => `${w.word},${w.count},${w.freq_pct}`).join("\n");
      _exportCsv(csv, `stats_rare_${r.label || "corpus"}.csv`);
    }
  } else if (_compareMode && _lastCompare) {
    const r = _lastCompare;
    const csv = `mot,n_${r.label_a},freq_${r.label_a},n_${r.label_b},freq_${r.label_b},ratio\n` +
      r.comparison.map(w => `${w.word},${w.count_a},${w.freq_a},${w.count_b},${w.freq_b},${w.ratio ?? ""}`).join("\n");
    _exportCsv(csv, `stats_compare_${r.label_a}_${r.label_b}.csv`);
  }
}

// ─── Main run function ────────────────────────────────────────────────────────

export async function runStats(): Promise<void> {
  if (!state.conn) return;

  const resultArea = document.getElementById("stats-results");
  if (!resultArea) return;
  resultArea.innerHTML = `<div class="stats-loading">Calcul en cours…</div>`;

  const runBtn = document.getElementById("stats-run-btn") as HTMLButtonElement | null;
  if (runBtn) runBtn.disabled = true;

  try {
    if (_compareMode) {
      const slotA = _readSlot("a");
      const slotB = _readSlot("b");
      const labelA = _readLabel("a");
      const labelB = _readLabel("b");
      const result = await fetchStatsCompare(state.conn, slotA, slotB, labelA, labelB);
      _lastCompare = result;
      _lastResult = null;
      resultArea.innerHTML = _renderCompareStats(result);
    } else {
      const slotA = _readSlot("a");
      const labelA = _readLabel("a");
      const result = await fetchLexicalStats(state.conn, slotA, labelA);
      _lastResult = result;
      _lastCompare = null;
      resultArea.innerHTML = _renderSingleStats(result);
    }

    // Wire export buttons
    resultArea.querySelectorAll<HTMLButtonElement>(".stats-export-btn").forEach(btn => {
      btn.addEventListener("click", () => _handleExport(btn.dataset.export ?? ""));
    });
  } catch (err) {
    resultArea.innerHTML = `<div class="stats-error">Erreur : ${_escHtml(err instanceof Error ? err.message : String(err))}</div>`;
  } finally {
    if (runBtn) runBtn.disabled = false;
  }
}

// ─── Panel builder ────────────────────────────────────────────────────────────

export function buildStatsPanel(): HTMLElement {
  const panel = elt("div", { class: "stats-panel hidden", id: "stats-panel" });

  const buildSlotGroup = (prefix: string, heading: string, extraClass = ""): HTMLElement => {
    const grp = elt("div", { class: `stats-slot-group ${extraClass}`, id: `stats-slot-${prefix}` });
    if (heading) grp.appendChild(elt("div", { class: "stats-slot-heading" }, heading));

    // Label
    const labelRow = elt("div", { class: "stats-filter-row" });
    labelRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-label` }, "Étiquette"));
    labelRow.appendChild(elt("input", { type: "text", class: "stats-filter-input", id: `stats-${prefix}-label`, placeholder: prefix === "a" ? "Corpus A" : "Corpus B" }));
    grp.appendChild(labelRow);

    // Language
    const langRow = elt("div", { class: "stats-filter-row" });
    langRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-lang` }, "Langue"));
    langRow.appendChild(elt("select", { class: "stats-filter-select", id: `stats-${prefix}-lang` }));
    grp.appendChild(langRow);

    // Role
    const roleRow = elt("div", { class: "stats-filter-row" });
    roleRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-role` }, "Rôle"));
    roleRow.appendChild(elt("select", { class: "stats-filter-select", id: `stats-${prefix}-role` }));
    grp.appendChild(roleRow);

    // Resource type
    const restypeRow = elt("div", { class: "stats-filter-row" });
    restypeRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-restype` }, "Type"));
    restypeRow.appendChild(elt("select", { class: "stats-filter-select", id: `stats-${prefix}-restype` }));
    grp.appendChild(restypeRow);

    // Family
    const famRow = elt("div", { class: "stats-filter-row" });
    famRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-family` }, "Famille"));
    famRow.appendChild(elt("select", { class: "stats-filter-select", id: `stats-${prefix}-family` }));
    grp.appendChild(famRow);

    // Doc IDs (manual)
    const docRow = elt("div", { class: "stats-filter-row" });
    docRow.appendChild(elt("label", { class: "stats-filter-label", for: `stats-${prefix}-doc-ids` }, "Doc IDs"));
    const docInput = elt("input", {
      type: "text",
      class: "stats-filter-input",
      id: `stats-${prefix}-doc-ids`,
      placeholder: "ex : 1,3,7",
      title: "IDs de documents séparés par des virgules (optionnel)",
    }) as HTMLInputElement;
    docRow.appendChild(docInput);
    grp.appendChild(docRow);

    return grp;
  };

  // Header with close button
  const header = elt("div", { class: "stats-panel-header" });
  header.appendChild(elt("span", { class: "stats-panel-title" }, "📊 Statistiques lexicales"));
  const closeBtn = elt("button", { class: "stats-close-btn", type: "button", title: "Fermer" }, "✕") as HTMLButtonElement;
  closeBtn.addEventListener("click", () => toggleStatsPanel(false));
  header.appendChild(closeBtn);
  panel.appendChild(header);

  // Controls row
  const controls = elt("div", { class: "stats-controls" });

  // Top N
  const topNRow = elt("div", { class: "stats-filter-row" });
  topNRow.appendChild(elt("label", { class: "stats-filter-label", for: "stats-top-n" }, "Top N"));
  const topNInput = elt("input", { type: "number", class: "stats-topn-input", id: "stats-top-n", value: "50", min: "5", max: "500" }) as HTMLInputElement;
  topNRow.appendChild(topNInput);
  controls.appendChild(topNRow);

  // Compare checkbox
  const compareRow = elt("div", { class: "stats-compare-row" });
  const compareCb = elt("input", { type: "checkbox", id: "stats-compare-cb" }) as HTMLInputElement;
  compareCb.addEventListener("change", _updateCompareModeUI);
  compareRow.appendChild(compareCb);
  compareRow.appendChild(elt("label", { for: "stats-compare-cb" }, "Mode Comparer A/B"));
  controls.appendChild(compareRow);

  // Run button
  const runBtn = elt("button", { class: "btn btn-primary stats-run-btn", id: "stats-run-btn", type: "button" }, "▶ Calculer") as HTMLButtonElement;
  runBtn.addEventListener("click", () => void runStats());
  controls.appendChild(runBtn);

  panel.appendChild(controls);

  // Slot A
  panel.appendChild(buildSlotGroup("a", ""));

  // Slot B (hidden by default)
  const slotB = buildSlotGroup("b", "Corpus B", "stats-slot-b-group");
  slotB.id = "stats-slot-b";
  slotB.style.display = "none";
  panel.appendChild(slotB);

  // Results area
  const resultsArea = elt("div", { class: "stats-results", id: "stats-results" });
  resultsArea.innerHTML = `<p class="stats-empty">Configurez les filtres et cliquez sur <strong>Calculer</strong>.</p>`;
  panel.appendChild(resultsArea);

  return panel;
}

// ─── Toggle ───────────────────────────────────────────────────────────────────

export function toggleStatsPanel(forceOpen?: boolean): void {
  const panel = document.getElementById("stats-panel");
  const btn = document.getElementById("stats-btn");
  if (!panel) return;

  const isOpen = !panel.classList.contains("hidden");
  const shouldOpen = forceOpen !== undefined ? forceOpen : !isOpen;

  if (shouldOpen) {
    panel.classList.remove("hidden");
    btn?.classList.add("active");
    populateStatsSelects();
  } else {
    panel.classList.add("hidden");
    btn?.classList.remove("active");
  }
}
